File: packages/agent-cache-py/betterdb_agent_cache/telemetry.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from opentelemetry import trace
from opentelemetry.trace import Tracer
from prometheus_client import REGISTRY as _DEFAULT_REGISTRY
from prometheus_client import CollectorRegistry, Counter, Gauge, Histogram

_OPERATION_BUCKETS = [0.001, 0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0]

# Module-level cache keyed by (id(registry), metric_name) to prevent
# duplicate-registration errors in multi-instance scenarios.
_metric_cache: dict[tuple[int, str], Any] = {}


def _get_or_create_counter(
    registry: CollectorRegistry,
    name: str,
    documentation: str,
    labelnames: list[str],
) -> Counter:
    key = (id(registry), name)
    if key not in _metric_cache:
        try:
            _metric_cache[key] = Counter(
                name, documentation, labelnames, registry=registry
            )
        except ValueError:
            existing = registry._names_to_collectors.get(name)
            if existing is None:
                raise
            _metric_cache[key] = existing
    return _metric_cache[key]  # type: ignore[return-value]


def _get_or_create_histogram(
    registry: CollectorRegistry,
    name: str,
    documentation: str,
    labelnames: list[str],
    buckets: list[float],
) -> Histogram:
    key = (id(registry), name)
    if key not in _metric_cache:
        try:
            _metric_cache[key] = Histogram(
                name, documentation, labelnames, buckets=buckets, registry=registry
            )
        except ValueError:
            existing = registry._names_to_collectors.get(name)
            if existing is None:
                raise
            _metric_cache[key] = existing
    return _metric_cache[key]  # type: ignore[return-value]


def _get_or_create_gauge(
    registry: CollectorRegistry,
    name: str,
    documentation: str,
    labelnames: list[str],
) -> Gauge:
    key = (id(registry), name)
    if key not in _metric_cache:
        try:
            _metric_cache[key] = Gauge(
                name, documentation, labelnames, registry=registry
            )
        except ValueError:
            existing = registry._names_to_collectors.get(name)
            if existing is None:
                raise
            _metric_cache[key] = existing
    return _metric_cache[key]  # type: ignore[return-value]


@dataclass
class AgentCacheMetrics:
    requests_total: Counter
    operation_duration: Histogram
    cost_saved: Counter
    stored_bytes: Counter
    active_sessions: Gauge


@dataclass
class Telemetry:
    tracer: Tracer
    metrics: AgentCacheMetrics


def create_telemetry(
    prefix: str,
    tracer_name: str,
    registry: CollectorRegistry | None = None,
) -> Telemetry:
    reg = registry or _DEFAULT_REGISTRY
    tracer = trace.get_tracer(tracer_name)

    metrics = AgentCacheMetrics(
        requests_total=_get_or_create_counter(
            reg,
            f"{prefix}_requests_total",
            "Total agent cache requests",
            ["cache_name", "tier", "result", "tool_name"],
        ),
        operation_duration=_get_or_create_histogram(
            reg,
            f"{prefix}_operation_duration_seconds",
            "Duration of agent cache operations in seconds",
            ["cache_name", "tier", "operation"],
            _OPERATION_BUCKETS,
        ),
        cost_saved=_get_or_create_counter(
            reg,
            f"{prefix}_cost_saved_total",
            "Estimated cost saved in dollars from cache hits",
            ["cache_name", "tier", "model", "tool_name"],
        ),
        stored_bytes=_get_or_create_counter(
            reg,
            f"{prefix}_stored_bytes_total",
            "Total bytes stored in cache",
            ["cache_name", "tier"],
        ),
        active_sessions=_get_or_create_gauge(
            reg,
            f"{prefix}_active_sessions",
            "Approximate number of active session threads",
            ["cache_name"],
        ),
    )

    return Telemetry(tracer=tracer, metrics=metrics)

File: packages/agent-cache/src/telemetry.ts
import { trace, type Tracer } from '@opentelemetry/api';
import {
  Counter,
  Histogram,
  Gauge,
  Registry,
  register as defaultRegistry,
  type CounterConfiguration,
  type HistogramConfiguration,
  type GaugeConfiguration,
} from 'prom-client';

interface TelemetryFactoryOptions {
  prefix: string;
  tracerName: string;
  registry?: Registry;
}

interface AgentCacheMetrics {
  requestsTotal: Counter;
  operationDuration: Histogram;
  costSaved: Counter;
  storedBytes: Counter;
  activeSessions: Gauge;
  discoveryWriteFailed: Counter;
}

export interface Telemetry {
  tracer: Tracer;
  metrics: AgentCacheMetrics;
}

function getOrCreateCounter(
  registry: Registry,
  config: CounterConfiguration<string>,
): Counter {
  const existing = registry.getSingleMetric(config.name);
  if (existing) return existing as Counter;
  return new Counter({ ...config, registers: [registry] });
}

function getOrCreateHistogram(
  registry: Registry,
  config: HistogramConfiguration<string>,
): Histogram {
  const existing = registry.getSingleMetric(config.name);
  if (existing) return existing as Histogram;
  return new Histogram({ ...config, registers: [registry] });
}

function getOrCreateGauge(
  registry: Registry,
  config: GaugeConfiguration<string>,
): Gauge {
  const existing = registry.getSingleMetric(config.name);
  if (existing) return existing as Gauge;
  return new Gauge({ ...config, registers: [registry] });
}

export function createTelemetry(opts: TelemetryFactoryOptions): Telemetry {
  const registry = opts.registry ?? defaultRegistry;
  const tracer = trace.getTracer(opts.tracerName);

  const operationBuckets = [0.001, 0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0];

  const requestsTotal = getOrCreateCounter(registry, {
    name: `${opts.prefix}_requests_total`,
    help: 'Total agent cache requests',
    labelNames: ['cache_name', 'tier', 'result', 'tool_name'],
  });

  const operationDuration = getOrCreateHistogram(registry, {
    name: `${opts.prefix}_operation_duration_seconds`,
    help: 'Duration of agent cache operations in seconds',
    labelNames: ['cache_name', 'tier', 'operation'],
    buckets: operationBuckets,
  });

  const costSaved = getOrCreateCounter(registry, {
    name: `${opts.prefix}_cost_saved_total`,
    help: 'Estimated cost saved in dollars from cache hits',
    labelNames: ['cache_name', 'tier', 'model', 'tool_name'],
  });

  const storedBytes = getOrCreateCounter(registry, {
    name: `${opts.prefix}_stored_bytes_total`,
    help: 'Total bytes stored in cache',
    labelNames: ['cache_name', 'tier'],
  });

  const activeSessions = getOrCreateGauge(registry, {
    name: `${opts.prefix}_active_sessions`,
    help: 'Approximate number of active session threads',
    labelNames: ['cache_name'],
  });

  const discoveryWriteFailed = getOrCreateCounter(registry, {
    name: `${opts.prefix}_discovery_write_failed_total`,
    help: 'Count of failed discovery-marker writes (best-effort HGET/HSET/SET operations against __betterdb:* keys)',
    labelNames: ['cache_name'],
  });

  return {
    tracer,
    metrics: {
      requestsTotal,
      operationDuration,
      costSaved,
      storedBytes,
      activeSessions,
      discoveryWriteFailed,
    },
  };
}
